include/ginkgo/core/preconditioner/batch_jacobi.hpp

```diff
@@ -0,0 +1,388 @@
+    friend class EnableBatchLinOp<Jacobi>;
+    friend class EnablePolymorphicObject<Jacobi, BatchLinOp>;
+
+public:
+    using EnableBatchLinOp<Jacobi>::convert_to;
+    using EnableBatchLinOp<Jacobi>::move_to;
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using matrix_type = batch::matrix::Csr<ValueType, IndexType>;
+
+    /**
+     * Returns the storage scheme used for storing Batched Jacobi blocks.
+     *
+     * @return the storage scheme used for storing Batched Jacobi blocks
+     *
+     */
+    const batched_jacobi_blocks_storage_scheme<index_type>&
+    get_blocks_storage_scheme() const noexcept
+    {
+        return blocks_storage_scheme_;
+    }
+
+    /**
+     *  Returns the block pointers.
+     *  @note Returns nullptr in case of a scalar jacobi preconditioner
+     * (max_block_size = 1).
+     *  @return the block pointers
+     */
+    const index_type* get_const_block_pointers() const noexcept
+    {
+        if (parameters_.max_block_size == 1) {
+            return nullptr;
+        }
+        return parameters_.block_pointers.get_const_data();
+    }
+
+    /**
+     *  Returns information about which blocks are the rows of the matrix part
+     * of.
+     *
+     *  @note Returns nullptr in case of a scalar jacobi preconditioner
+     * (max_block_size = 1).
+     *
+     */
+    const index_type* get_const_row_is_part_of_which_block_info() const noexcept
+    {
+        if (parameters_.max_block_size == 1) {
+            return nullptr;
+        }
+        return row_part_of_which_block_info_.get_const_data();
+    }
+
+    /**
+     *  Returns the cumulative blocks storage array
+     *
+     *  @note Returns nullptr in case of a scalar jacobi preconditioner
+     * (max_block_size = 1).
+     *
+     */
+    const index_type* get_const_blocks_cumulative_storage() const noexcept
+    {
+        if (parameters_.max_block_size == 1) {
+            return nullptr;
+        }
+        return blocks_cumulative_storage_.get_const_data();
+    }
+
+    /**
+     * Returns the max block size.
+     *
+     * @return the max block size
+     */
+    uint32 get_max_block_size() const noexcept
+    {
+        return parameters_.max_block_size;
+    }
+
+    /**
+     * Returns the number of blocks in an individual batch entry.
+     *
+     * @return the number of blocks in an individual batch entry.
+     *
+     */
+    size_type get_num_blocks() const noexcept { return num_blocks_; }
+
+
+    /**
+     * Returns the pointer to the memory used for storing the block data.
+     *
+     * Element (`i`, `j`) of the block, which belongs to the batch entry with
+     * index = batch_id and has local id = "block_id" within its batch entry is
+     * stored at the address = get_const_blocks() +
+     * storage_scheme.get_global_block_offset(batch_id, num_blocks, block_id,
+     * cumulative_blocks_storage) + i * storage_scheme.get_stride(block_id,
+     * block_pointers) + j
```

Comment:
This seems to be an issue again with the storage scheme. Why isn't it just `address = storage_scheme(batch_id, block_id, i, j)`?

Reply:
This is not necessary in the kernels, but is just a note to document how things are stored.